### PR TITLE
sampling: add Top-nσ sampler

### LINKED ISSUE #11057: Feature Request: Top-nσ sampler
### Prerequisites

- [X] I am running the latest code. Mention the version if possible as well.
- [X] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [X] I searched using keywords relevant to my issue to make sure that I am creating a new issue that is not already open (or closed).
- [X] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new and useful enhancement to share.

### Feature Description

Here are the key points:

**The Problem:** When LLMs generate text, they typically use either greedy decoding (always picking the most likely token) or temperature sampling. Current sampling methods often struggle to balance diversity with accuracy, especially for reasoning tasks.

**The Innovation:** The authors discovered that when LLMs generate tokens, the logits (pre-softmax scores) naturally separate into two regions:

- A "noisy" region following a Gaussian distribution (background noise)

- An "informative" region containing the actually relevant tokens

**The Solution:** Top-nσ works by:

- Identifying the maximum logit value

- Selecting tokens that are within n standard deviations (σ) of this maximum

- Only sampling from these selected tokens

- Using temperature to control sampling within this filtered set

**Key Benefits:**

- Maintains consistent performance even at high temperatures, unlike other methods

- Computationally efficient as it operates directly on logits

- Outperforms both existing sampling methods and greedy decoding on reasoning tasks

- Works particularly well for tasks requiring careful reasoning

**Results:** The method was tested on four reasoning-focused datasets and showed superior performance, especially at higher temperatures where other methods typically fail.

The paper essentially shows that by being more selective about which tokens to sample from based on their statistical properties, you can get better and more reliable results from language models, particularly for tasks that require careful reasoning.

### Motivation

Looks to be the best sampler yet, and will be a clear differentiator for llama.cpp

### Possible Implementation

See white paper: _"Top-nσ Not All Logits Are You Need"_

### PR DESCRIPTION
**Pull request by VJHack**
### Top-nσ: Not All Logits Are You Need
https://arxiv.org/pdf/2411.07641
The authors of this paper propose a new sampling method known as Top-nσ. The main feature of this sampler is that "unlike existing methods (e.g., top-p, min-p) that inadvertently include more noise tokens at higher temperatures, top-nσ maintains a stable sampling space regardless of temperature scaling". They discovered that logits natually separate into a gaussian-distributed noisy region and an informative region. 

This PR implements the sampling method proposed in the paper. Here the algorithm implemented from the paper:
Screen Shot 2025-01-13 at 6 14 41 PM

Since the manipulation is done directly on the logits pre-softmax, I added it as a stand-alone sampler instead of chaining it with the common samplers. The changes only add support for `llama-cli`.
`sampler chain: logits -> logit-bias -> temp -> top-n-sigma -> dist `

I'm aware that this algorithm is[ still in it's early phases](https://github.com/Tomorrowdawn/top_nsigma) so we could tag this as demo for now but I'll leave that choice up to the maintainers.

resolves #11057 

Relavent Links:
https://huggingface.co/papers/2411.07641
https://arxiv.org/pdf/2411.07641
https://github.com/Tomorrowdawn/top_nsigma
https://github.com/ggerganov/llama.cpp/issues/11057

### CODE REVIEW

**Comment by MaggotHATE** [comment]
Thank you for this implementation! Top-nσ is definitely special and needs a lot of testing.

I like the results so far, especially since high temperature is not a problem, as shown in the paper, and I'm going to test it more and see what its limitations are.

File: common/sampling.cpp [hunk 1]
```diff
@@ -167,45 +167,50 @@ struct common_sampler * common_sampler_init(const struct llama_model * model, co
...
-                case COMMON_SAMPLER_TYPE_DRY:
-                    {
-                        std::vector<const char *> c_breakers;
-                        c_breakers.reserve(params.dry_sequence_breakers.size());
-                        for (const auto & str : params.dry_sequence_breakers) {
-                            c_breakers.push_back(str.c_str());
+        if(params.top_n_sigma >= 0) {
+            llama_sampler_chain_add(result->chain, llama_sampler_init_temp(params.temp));
+            llama_sampler_chain_add(result->chain, llama_sampler_init_top_n_sigma(params.top_n_sigma));
+        } else {
```
Comment (slaren) [inline common/sampling.cpp:173]: I am not convinced that this is desirable. Using top-k before this sampler should improve performance significantly, and the difference in output is likely to be negligible.

Comment (VJHack) [reply common/sampling.cpp]: Thanks for pointing this out! I see your point that processing the entire vocabulary to apply top-nσ sampling is computationally expensive, and the negligible difference in output by chaining this after top-k doesn't justify the cost.

I made the necessary change to apply top-k before this to improve performance that way we're processing fewer tokens to achieve similar results. 
`sampler chain: logits -> logit-bias -> top-k -> temp -> top-n-sigma -> dist `

The main benefit we should be getting from this sampler is a stable sampling space regardless of temperature scaling. By chaining it after top-k, we can now control the size of this sampling space too.

Comment (MaggotHATE) [reply common/sampling.cpp]: Moreover, I'm currently stuck with two models that almost require some form of repetition penalty (either a traditional one, or DRY), and the current implementation doesn't allow that.

I get that this sampler wasn't designed with creative writing in mind, but it can be used for that purpose. Mistral Small 22B show especially good results, but some finetuned models need something to battle repetitions.

FYI I've also tested introducing (gaussian) noise as a sampler before `top_n_sigma`, but it broke horribly. Seems like it's quite sensitive to that.

Comment (Tomorrowdawn) [reply common/sampling.cpp]: Thank you for testing our algorithm! Indeed, this sampler is just meant to eliminate the Gaussian noise, but we're actually seeing the same thing in our ongoing research. We find it works quite well for creative writing, and strangely enough, it performs really well at low temperatures. 
image

Comment (MaggotHATE) [reply common/sampling.cpp]: @Tomorrowdawn Hi! Thank you for this research!

At the moment I see the following in regards to repetitions:
1. Traditional repetition penalty (logits -> bias -> top_k -> repetition_penalty -> temp -> top_n_sigma) helps, but it needs more careful testing and tuning - sometimes it can steer creativity into broken logics.
2. DRY works really well, but at the moment I've only tested it in a legacy order (i.e. logits -> bias -> DRY -> top_k -> temp -> top_n_sigma). Ideally, it should go after top_k for optimization, but so far it's quite good.

In regards to gaussian noise, it was used [before](https://github.com/kalomaze/koboldcpp/releases/tag/noisy-v1) as an experiment to improve "creativity" and battle repetitions. I've been using it for a long time now, and it seems to be effective - but it's still not that useful against repetitions as DRY, for example. At the same time, it works well with other existing samplers (both as a separate sampler and as a part of `p_step`, for example).

As for high temperatures, I've tested `top_n_sigma` with temp = 5 previously, and it was still good enough, but the prompt adherence started to suffer a bit, which is a problem for creative writing (especially with longer prompts). In general, I found `top_n_sigma` to be more random (less controllable?) in its own way: the length of responses and overall creativity seem to be different from time to time on the same parameters.

For example, I was testing it with `temp` = 1.9, and results ranged from low-temp-like to long ones. The "confidence" (just an experiment, I measure it as simple as "percentage of chosen candidates with the chance of 50% and higher") was all over the place too, from 0.8 to 0.4.

I'm still testing, and I will look more into low temperatures with `top_n_sigma`. This ampler is very interesting, and the results are definitely unique.

Comment (Tomorrowdawn) [reply common/sampling.cpp]: Thank you for your thorough testing! 

For the noisy sampling:  
Since `top_n_sigma` is fundamentally sensitive to logits distribution, I think a better approach would be to apply `top_n_sigma` (with a slightly larger `nsigma`) before introducing noise, as the noise-induced variance would significantly interfere with `top_n_sigma`'s effectiveness. 

For repetition and the inconsistent performance: 
I'm not sure whether this behavior should be considered 'good' or 'bad', as `top_n_sigma` is designed to be general-purpose and to reflect the model's inherent capabilities rather than introducing human priors.   Specialized sampling algorithms should be more suitable for this specific aspect. 

Your experiments have been truly insightful. I believe `top_n_sigma`'s main strength lies in providing a safety net for the word space (same as VJHack's comment), while the probability distribution itself might still require specialized samplers.

Comment (MaggotHATE) [reply common/sampling.cpp]: > Since `top_n_sigma` is fundamentally sensitive to logits distribution, I think a better approach would be to apply `top_n_sigma` (with a slightly larger `nsigma`) before introducing noise, as the noise-induced variance would significantly interfere with `top_n_sigma`'s effectiveness.

Interesting, thank you, I'll test it. On the "a slightly larger `nsigma`" - I saw hallucinations with `n_sigma` set to 2, unfortunately, even with low temperature.

> Specialized sampling algorithms should be more suitable for this specific aspect.

The question is, can (should?) `top_n_sigma` be combined with other samplers in the common chain? If I read the paper right, it was designed as an answer to limitations of `top-p`, `min-p` and other similar samplers. Maybe using `top_n_sigma` in place of these samplers and combining it in a more flexible way would give even better results.

Comment (Tomorrowdawn) [reply common/sampling.cpp]: About the hyperparameter: `top_n_sigma` is equivalent to `min_p` where p = 1/exp(nsigma). Take sigma=2.2 as a typical value and you will get n=1 for p = 0.1, and n=2 for p=0.01. It doesn't scale linearly. 1.3~1.5 might better match what I meant by 'slightly larger' (sorry for not being clear)... As a default option, 1.0 is a simple, memorable, and practical value.

Regarding combining with other samplers: Since it needs to calculate standard deviation, and it's designed to eliminate noise from the original distribution, `top_nsigma` is typically used as the [first sampler](https://github.com/aphrodite-engine/aphrodite-engine/pull/825#issuecomment-2487247415)

Comment (MaggotHATE) [reply common/sampling.cpp]: > 1.3~1.5 might better match what I meant by 'slightly larger' (sorry for not being clear)...

Thanks for clarification! In such case, `top_n_sigma` shouldn't be an integer variable as it is right now in this PR.

> top_nsigma is typically used as the first sampler

Hmm, what about top-k with high (1000, for example) values? If both `temp` and `top_n_sigma` will be working on untruncated list of candidates, it might decrease performance a bit.

In practice, unless `p-step` is used, the number of candidates that are kept till `dist` is less than 1000 (in most cases it's less than 200 even). Having `top-k` as the first sampler would help performance even at such high value - but would it interfere with `top_n_sigma` in such a case?

Comment (Tomorrowdawn) [reply common/sampling.cpp]: Oh, I see the point. The situation is that the code we're discussing is different: typically, `transformers` or `vllm` users would pass the **complete** intermediate logits vector (since these operations are fast on GPU), and `top_n_sigma` would crash due to negative infinities introduced by other samplers. If we only keep the candidates selected by top-k, this is completely feasible. The previously mentioned chain of `top-k->temp->top_nsigma` is correct.

Comment (VJHack) [reply common/sampling.cpp]: > `top_n_sigma` shouldn't be an integer variable as it is right now in this PR.
@MaggotHATE I converted top_n_sigma to a float. This seams to give us much more control over the sampling space rather than leaving it as an integer. 

While it was an int, I noticed that any `top_n_sigma` greater than 2 had little to no effect on the output, leaving us to chose between only 3 values for `top_n_sigma` (0,1, and 2). Thanks for the suggestion!

Comment (MaggotHATE) [reply common/sampling.cpp]: @VJHack Thank you! I'm currently testing with `top_n_sigma` = 1.3, and it works quite well!

> top_n_sigma is designed to be general-purpose and to reflect the model's inherent capabilities rather than introducing human priors.

@Tomorrowdawn you were right about this, and even more: it looks like `top_n_sigma` can increase sensitivity to the prompt.

For example, I was testing with a prompt formatted using colons (i.e. "Title:..."). A finetuned model that had such formatting in the finetuning dataset worked correctly - until I used the same formatting for something that was not in the dataset (i.e. "TRIVIA:..."). At that point the quality of responses degraded significantly. Looks like my previous complains about repetitions stemmed from this too (although, repeated words may still appear close to each other in the text). I've never seen such a strong effect with other sampling algorithms.

It's a very interesting result, because it means that `top_n_sigma` allows us to focus on testing the "quality" of finetuning, so to speak. None of the problems mentioned above happened with the foundation models (Mistral Small 22B in this case, but I've also tested Nemo 12B).

**Review by slaren — APPROVED** [review]

**Comment by hdu-hh** [comment]
Sorry for the ping, but what is needed so that this approved PR gets merged? I like the new feature very much as it allows the benefits of a higher temperature without the usual drawbacks.

**Comment by VJHack** [comment]
@hdu-hh I agree! This would be a cool feature to have but I don't think it's a priority for the maintainers at the moment.
They'll get around to it if they see it benefiting the community. This isn't one of the mainstream sampling algos so it's understandable if they don't accept it. We should label it as demo at the very least.

File: src/llama-sampling.cpp [hunk 2]
```diff
@@ -1648,6 +1648,73 @@ struct llama_sampler * llama_sampler_init_penalties(
...
+
+struct llama_sampler * llama_sampler_init_top_n_sigma(float n) {
+    return new llama_sampler {
+        /* .iface = */ &llama_sampler_top_n_sigma_i,
+        /* .ctx   = */ new llama_sampler_top_n_sigma {
+            /* .n = */ n,
+                        },
+    };
+}
+
```
Comment (ggerganov) [inline src/llama-sampling.cpp:1717]: Use the `llama_sampler_init()` instead of `new` (see #11727).

Fix formatting:

```suggestion
struct llama_sampler * llama_sampler_init_top_n_sigma(float n) {
    return new llama_sampler {
        /* .iface = */ &llama_sampler_top_n_sigma_i,
        /* .ctx   = */ new llama_sampler_top_n_sigma {
            /* .n = */ n,
        },
    };
}

```

File: src/llama-sampling.cpp [hunk 3]
```diff
@@ -1648,6 +1648,73 @@ struct llama_sampler * llama_sampler_init_penalties(
...
+static void llama_sampler_top_n_sigma_apply(struct llama_sampler * smpl, llama_token_data_array * cur_p) {
+    const auto * ctx = (llama_sampler_top_n_sigma *) smpl->ctx;
+
+    // find max logit and calculate mean
+    float max = cur_p->data[0].logit;
+    float logits_sum = 0;
+    for (size_t i = 0; i < cur_p->size; ++i) {
+        if(cur_p->data[i].logit > max){
```
Comment (ggerganov) [inline src/llama-sampling.cpp:1668]:
```suggestion
        if (cur_p->data[i].logit > max) {
```

File: src/llama-sampling.cpp [hunk 4]
```diff
@@ -1648,6 +1648,73 @@ struct llama_sampler * llama_sampler_init_penalties(
...
+            max = cur_p->data[i].logit;
+        }
+        logits_sum += cur_p->data[i].logit;
+    }
+    float mean = logits_sum/cur_p->size;
+
+    // calculate standard deviation
+    float acc = 0;
+    for(size_t i = 0; i < cur_p->size; ++i){
```
Comment (ggerganov) [inline src/llama-sampling.cpp:1677]:
```suggestion
    for (size_t i = 0; i < cur_p->size; ++i) {
```

File: src/llama-sampling.cpp [hunk 5]
```diff
@@ -1648,6 +1648,73 @@ struct llama_sampler * llama_sampler_init_penalties(
...
+        acc += pow(cur_p->data[i].logit - mean, 2);
+    }
+    float std = sqrt(acc/cur_p->size);
+
+    //apply mask
+    for(size_t i = 0; i < cur_p->size; ++i){
```
Comment (ggerganov) [inline src/llama-sampling.cpp:1683]:
```suggestion
    for (size_t i = 0; i < cur_p->size; ++i) {
```

File: src/llama-sampling.cpp [hunk 6]
```diff
@@ -1648,6 +1648,73 @@ struct llama_sampler * llama_sampler_init_penalties(
...
+        if(cur_p->data[i].logit < max - (ctx->n * std)) {
```
Comment (ggerganov) [inline src/llama-sampling.cpp:1684]:
```suggestion
        if (cur_p->data[i].logit < max - (ctx->n * std)) {
```

File: common/sampling.cpp [hunk 7]
```diff
@@ -167,45 +167,51 @@ struct common_sampler * common_sampler_init(const struct llama_model * model, co
...
     if (params.mirostat == 0) {
-        for (const auto & cnstr : params.samplers) {
-            switch (cnstr) {
-                case COMMON_SAMPLER_TYPE_DRY:
-                    {
-                        std::vector<const char *> c_breakers;
-                        c_breakers.reserve(params.dry_sequence_breakers.size());
-                        for (const auto & str : params.dry_sequence_breakers) {
-                            c_breakers.push_back(str.c_str());
+        if(params.top_n_sigma >= 0) {
```
Comment (ggerganov) [inline common/sampling.cpp:170]:
```suggestion
        if (params.top_n_sigma >= 0) {
```

File: src/llama-sampling.cpp [hunk 8]
```diff
@@ -1648,6 +1648,73 @@ struct llama_sampler * llama_sampler_init_penalties(
...
+            cur_p->data[i].logit = -INFINITY;
+        }
+    }
+    llama_sampler_softmax_impl(cur_p);
+}
+
+static struct llama_sampler * llama_sampler_top_n_sigma_clone(const struct llama_sampler * smpl){
```
Comment (ggerganov) [inline src/llama-sampling.cpp:1691]:
```suggestion
static struct llama_sampler * llama_sampler_top_n_sigma_clone(const struct llama_sampler * smpl) {
```

**Review by ggerganov — APPROVED** [review]

File: src/llama-sampling.cpp (hunk 2 above)
Comment (VJHack) [reply src/llama-sampling.cpp]: I made the change to `llama_sampler_init()` as done with the other samplers.

**Comment by SmallAndSoft** [comment]
We are just a small step away from bringing it to users: https://github.com/ggml-org/llama.cpp/discussions/14637
